src/nupic/support/group_by.py

```diff
@@ -0,0 +1,161 @@
+  @param fn  (function) the key function to perform the groupings
+
+  @return (tuple) A (key, GroupByGenerator) tuple where the first value is the
+                  key of the group, and second value is the generator that
+                  generates the values in the group.
+  """
+  length = 1
+  beginning = 0
+  l = len(lis)
+
+  if l > 0:
+    key = fn(lis[0])
+    for i in xrange(1, len(lis)):
+      val = fn(lis[i])
+
+      if val != key:
+        yield (key, GroupByGenerator(lis, beginning, length))
+        length = 1
+        beginning = i
+        key = val
+      else:
+        length += 1
+
+  if l > 0: # yield last group
+    yield (key, GroupByGenerator(lis, beginning, l - beginning))
+
+
+def groupByN(*args):
+  """ A utility function wrapper based on groupby in itertools. Allows to
+      walk across n sorted lists with respect to their key functions
+      and yields a tuple of n lists of the members of the next *smallest*
+      group.
+
+  @param args (list) a list of arguments alternating between sorted lists and
+                       their respective key functions. The lists should be
+                       sorted with respect to their key function.
+
+  @return (tuple) a n + 1 dimensional tuple, where the first element is the
+                    key of the group and the other n entries are lists of
+                    objects that are a member of the current group that is being
+                    iterated over in the nth list passed in. Note that this
+                    is a generator and a n+1 dimensional tuple is yielded for
+                    every group.
+
+  Notes: Read up on groupby here:
+         https://docs.python.org/dev/library/itertools.html#itertools.groupby
+
+         itertools.groupby was not used and the justification can be seen
+         here https://github.com/numenta/nupic/pull/3254#discussion_r74164564
+
+"""
+  generatorList = [] # list of each list's (k, group) tuples
+
+  if len(args) % 2 == 1:
+    raise ValueError("Must have a key function for every list.")
+
+  # populate above lists
+  for i in xrange(0, len(args), 2):
+    listn = args[i]
+    fn = args[i + 1]
+    generatorList.append(groupby(listn, fn))
+
+  n = len(generatorList)
+
+  advanceList = [True] * n # start by advancing everyone.
+  nextList = [None] * n
+  # while all lists aren't exhausted walk through each group in order
+  while True:
+    for i in xrange(n):
+      if advanceList[i]:
+        try:
+          nextList[i] = generatorList[i].next()
+        except StopIteration:
+          nextList[i] = None
+
+    # no more values to process in any of the generators
+    if all([entry is None for entry in nextList]):
+      break
+
+    # find the smallest next group and all lists that have an element in it
+    argIndices = [] # array of indices corresponding to the lists that have at
+                    # least one element that is a member of the minKeyVal group
+    minKeyVal = float("inf")
```

Comment:
This argIndices list is kind of hard to read.

Instead, maybe do

~~~python
minKeyVal = min(nextVal[0] for nextVal in nextList
                if nextVal is not None)
~~~

Then further down just check

~~~python
for i in xrange(n):
  if nextList[i] is not None and nextList[i][0] == minKeyVal:
    ...
  else:
    ...
~~~

Do you see any downsides with that approach?